--- caffe/include/caffe/util/io.hpp	2017-01-31 23:29:02.000000000 +0800
+++ ../include/caffe/util/io.hpp	2017-01-30 15:43:29.000000000 +0800
@@ -1,7 +1,9 @@
 #ifndef CAFFE_UTIL_IO_H_
 #define CAFFE_UTIL_IO_H_
 
+#ifndef DISABLE_BOOST
 #include <boost/filesystem.hpp>
+#endif
 #include <iomanip>
 #include <iostream>  // NOLINT(readability/streams)
 #include <string>
@@ -19,6 +21,7 @@
 namespace caffe {
 
 using ::google::protobuf::Message;
+#ifndef DISABLE_BOOST
 using ::boost::filesystem::path;
 
 inline void MakeTempDir(string* temp_dirname) {
@@ -48,6 +51,7 @@
   *temp_filename =
     (temp_files_subpath/caffe::format_int(next_temp_file++, 9)).string();
 }
+#endif
 
 bool ReadProtoFromTextFile(const char* filename, Message* proto);
 
@@ -129,6 +133,14 @@
 bool DecodeDatumNative(Datum* datum);
 bool DecodeDatum(Datum* datum, bool is_color);
 
+#ifdef CAFFE_COMPACT
+void ReadNetParamsFromTextFileOrDie(const string& param_file,
+                                    NetParameter* param);
+
+void ReadNetParamsFromBinaryFileOrDie(const string& param_file,
+                                      NetParameter* param);
+#endif
+
 #ifdef USE_OPENCV
 cv::Mat ReadImageToCVMat(const string& filename,
     const int height, const int width, const bool is_color);
