--- caffe/include/caffe/util/math_functions.hpp	2017-04-19 19:08:25.000000000 +0800
+++ ../include/caffe/util/math_functions.hpp	2017-04-19 17:09:45.000000000 +0800
@@ -4,7 +4,12 @@
 #include <stdint.h>
 #include <cmath>  // for std::fabs and std::signbit
 
+#ifdef USE_GLOG
 #include "glog/logging.h"
+#endif
+#ifndef NO_CAFFE_MOBILE
+#include <string.h>
+#endif
 
 #include "caffe/common.hpp"
 #include "caffe/util/device_alternate.hpp"
