--- caffe/include/caffe/caffe.hpp	2017-01-27 09:51:55.335623700 +0800
+++ ../include/caffe/caffe.hpp	2017-02-04 22:35:34.594153197 +0800
@@ -10,10 +10,14 @@
 #include "caffe/layer.hpp"
 #include "caffe/layer_factory.hpp"
 #include "caffe/net.hpp"
+#ifdef NO_CAFFE_MOBILE
 #include "caffe/parallel.hpp"
+#endif
 #include "caffe/proto/caffe.pb.h"
+#ifdef NO_CAFFE_MOBILE
 #include "caffe/solver.hpp"
 #include "caffe/solver_factory.hpp"
+#endif
 #include "caffe/util/benchmark.hpp"
 #include "caffe/util/io.hpp"
 #include "caffe/util/upgrade_proto.hpp"
