--- caffe/include/caffe/common.hpp	2017-04-19 19:07:36.000000000 +0800
+++ ../include/caffe/common.hpp	2017-04-19 17:09:45.000000000 +0800
@@ -1,9 +1,19 @@
 #ifndef CAFFE_COMMON_HPP_
 #define CAFFE_COMMON_HPP_
 
+#ifdef USE_BOOST
 #include <boost/shared_ptr.hpp>
+#else
+#include <memory>
+#endif
+#ifdef NO_CAFFE_MOBILE
 #include <gflags/gflags.h>
+#endif
+#ifdef USE_GLOG
 #include <glog/logging.h>
+#else
+#include "caffe/glog_wrapper.hpp"
+#endif
 
 #include <climits>
 #include <cmath>
@@ -22,6 +32,7 @@
 #define STRINGIFY(m) #m
 #define AS_STRING(m) STRINGIFY(m)
 
+#ifdef NO_CAFFE_MOBILE
 // gflags 2.1 issue: namespace google was changed to gflags without warning.
 // Luckily we will be able to use GFLAGS_GFLAGS_H_ to detect if it is version
 // 2.1. If yes, we will add a temporary solution to redirect the namespace.
@@ -30,6 +41,7 @@
 #ifndef GFLAGS_GFLAGS_H_
 namespace gflags = google;
 #endif  // GFLAGS_GFLAGS_H_
+#endif  // NO_CAFFE_MOBILE
 
 // Disable the copy and assignment operator for a class.
 #define DISABLE_COPY_AND_ASSIGN(classname) \
@@ -74,9 +86,13 @@
 
 namespace caffe {
 
+#ifdef USE_BOOST
 // We will use the boost shared_ptr instead of the new C++11 one mainly
 // because cuda does not work (at least now) well with C++11 features.
 using boost::shared_ptr;
+#else
+using std::shared_ptr;
+#endif
 
 // Common functions and classes from std that caffe often uses.
 using std::fstream;
