--- caffe/include/caffe/filler.hpp	2017-01-31 23:29:02.000000000 +0800
+++ ../include/caffe/filler.hpp	2017-01-30 15:43:29.000000000 +0800
@@ -45,6 +45,7 @@
   }
 };
 
+#ifndef CAFFE_COMPACT
 /// @brief Fills a Blob with uniformly distributed values @f$ x\sim U(a, b) @f$.
 template <typename Dtype>
 class UniformFiller : public Filler<Dtype> {
@@ -260,6 +261,7 @@
          << "Sparsity not supported by this Filler.";
   }
 };
+#endif
 
 /**
  * @brief Get a specific filler from the specification given in FillerParameter.
@@ -272,6 +274,7 @@
   const std::string& type = param.type();
   if (type == "constant") {
     return new ConstantFiller<Dtype>(param);
+#ifndef CAFFE_COMPACT
   } else if (type == "gaussian") {
     return new GaussianFiller<Dtype>(param);
   } else if (type == "positive_unitball") {
@@ -284,6 +287,7 @@
     return new MSRAFiller<Dtype>(param);
   } else if (type == "bilinear") {
     return new BilinearFiller<Dtype>(param);
+#endif
   } else {
     CHECK(false) << "Unknown filler name: " << param.type();
   }
