--- caffe/src/caffe/layers/base_data_layer.cpp	2017-04-19 19:07:37.000000000 +0800
+++ ../src/caffe/layers/base_data_layer.cpp	2017-04-19 17:09:45.000000000 +0800
@@ -1,9 +1,13 @@
+#ifdef USE_BOOST
 #include <boost/thread.hpp>
+#endif
 #include <vector>
 
 #include "caffe/blob.hpp"
 #include "caffe/data_transformer.hpp"
+#ifdef USE_BOOST
 #include "caffe/internal_thread.hpp"
+#endif
 #include "caffe/layer.hpp"
 #include "caffe/layers/base_data_layer.hpp"
 #include "caffe/proto/caffe.pb.h"
@@ -32,6 +36,7 @@
   DataLayerSetUp(bottom, top);
 }
 
+#ifdef NO_CAFFE_MOBILE
 template <typename Dtype>
 BasePrefetchingDataLayer<Dtype>::BasePrefetchingDataLayer(
     const LayerParameter& param)
@@ -129,8 +134,11 @@
 #ifdef CPU_ONLY
 STUB_GPU_FORWARD(BasePrefetchingDataLayer, Forward);
 #endif
+#endif // NO_CAFFE_MOBILE
 
 INSTANTIATE_CLASS(BaseDataLayer);
+#ifdef NO_CAFFE_MOBILE
 INSTANTIATE_CLASS(BasePrefetchingDataLayer);
+#endif // NO_CAFFE_MOBILE
 
 }  // namespace caffe
