--- caffe/src/caffe/layers/base_data_layer.cpp	2017-01-27 09:51:56.330123800 +0800
+++ ../src/caffe/layers/base_data_layer.cpp	2017-02-04 22:35:34.654099433 +0800
@@ -1,9 +1,13 @@
+#ifdef USE_BOOST
 #include <boost/thread.hpp>
+#endif
 #include <vector>
 
 #include "caffe/blob.hpp"
 #include "caffe/data_transformer.hpp"
+#ifdef USE_BOOST
 #include "caffe/internal_thread.hpp"
+#endif
 #include "caffe/layer.hpp"
 #include "caffe/layers/base_data_layer.hpp"
 #include "caffe/proto/caffe.pb.h"
@@ -13,8 +17,12 @@
 
 template <typename Dtype>
 BaseDataLayer<Dtype>::BaseDataLayer(const LayerParameter& param)
+#ifdef NO_CAFFE_MOBILE
     : Layer<Dtype>(param),
       transform_param_(param.transform_param()) {
+#else
+    : Layer<Dtype>(param) {
+#endif
 }
 
 template <typename Dtype>
@@ -25,13 +33,16 @@
   } else {
     output_labels_ = true;
   }
+#ifdef NO_CAFFE_MOBILE
   data_transformer_.reset(
       new DataTransformer<Dtype>(transform_param_, this->phase_));
   data_transformer_->InitRand();
+#endif
   // The subclasses should setup the size of bottom and top
   DataLayerSetUp(bottom, top);
 }
 
+#ifdef NO_CAFFE_MOBILE
 template <typename Dtype>
 BasePrefetchingDataLayer<Dtype>::BasePrefetchingDataLayer(
     const LayerParameter& param)
@@ -129,8 +140,11 @@
 #ifdef CPU_ONLY
 STUB_GPU_FORWARD(BasePrefetchingDataLayer, Forward);
 #endif
+#endif
 
 INSTANTIATE_CLASS(BaseDataLayer);
+#ifdef NO_CAFFE_MOBILE
 INSTANTIATE_CLASS(BasePrefetchingDataLayer);
+#endif
 
 }  // namespace caffe
